certora/specs/Splits.spec

```diff
@@ -0,0 +1,418 @@
+
+    // calling the split() for userA over assertId1
+    uint128 userA_collectableAmtAssetId1; uint128 userA_splitAmtAssetId1;
+    userA_collectableAmtAssetId1, userA_splitAmtAssetId1 = split(e, userA_Id, assetId1, true);
+
+    // recording the state after split()
+    uint128 userA_splittableAssetId1_After = splittable(e, userA_Id, assetId1);
+    uint128 userA_collectableAssetId1_After = collectable(e, userA_Id, assetId1);
+    uint128 userA_splittableAssetId2_After = splittable(e, userA_Id, assetId2);
+    uint128 userA_collectableAssetId2_After = collectable(e, userA_Id, assetId2);
+
+    uint128 userB_splittableAssetId1_After = splittable(e, userB_Id, assetId1);
+    uint128 userB_collectableAssetId1_After = collectable(e, userB_Id, assetId1);
+    uint128 userB_splittableAssetId2_After = splittable(e, userB_Id, assetId2);
+    uint128 userB_collectableAssetId2_After = collectable(e, userB_Id, assetId2);
+
+    // the expectation:
+    // splittable and collectable for any user's assetId2 stays the same
+    assert userA_splittableAssetId2_After == userA_splittableAssetId2_Before;
+    assert userA_collectableAssetId2_After == userA_collectableAssetId2_Before;
+
+    assert userB_splittableAssetId2_After == userB_splittableAssetId2_Before;
+    assert userB_collectableAssetId2_After == userB_collectableAssetId2_Before;
+}
+
+
+/// @notice Money is not lost in the system when split() is called:
+/// UserA has splittable balance and one splits receiver - userB,
+/// then split() is called on userA
+/// We verify that the sum (splittable + collectable) of (userA + userB)
+/// are invariant of the split
+rule moneyNotLostDuringSplit() {
+    env e; uint256 assetId;
+    uint256 userA_Id; uint256 userB_Id;
+    require userA_Id != userB_Id;
+
+    // recording the state before split()
+    uint128 userA_splittableBefore = splittable(e, userA_Id, assetId);
+    require userA_splittableBefore > 0;
+    uint128 userA_collectableBefore = collectable(e, userA_Id, assetId);
+    uint128 userB_splittableBefore = splittable(e, userB_Id, assetId);
+    uint128 userB_collectableBefore = collectable(e, userB_Id, assetId);
+
+    // setting up the currReceivers[] of userA
+    uint256 length; uint256 index1; uint256 userId1; uint32 weight1;
+    length = getCurrSplitsReceiverLocaLength(e, true);
+    require length == 1;
+    userId1, weight1 = getCurrSplitsReceiverLocalArr(e, true, index1);
+    //require index1 == 0;
+    require userId1 == userB_Id;
+    require weight1 <= 1000000;  // safe assumptions since the function _assertSplitsValid()
+    // verified that totalWeight <= _TOTAL_SPLITS_WEIGHT upon _setSplits()
+    // we required that there is only one receiver, therefore totalWeight = weight1
+    // _TOTAL_SPLITS_WEIGHT == 1000000, hence the assumption above is safe
+
+    // calling the split() on userA
+    uint128 userA_collectableAmt;   uint128 userA_splitAmt;
+    userA_collectableAmt, userA_splitAmt = split(e, userA_Id, assetId, true);
```

Comment:
Here, if I understand correctly, you split to another user different than userB.
Why do you exclude the case where you call split to userB also?

Reply:
In rule moneyNotLostDuringSplit, the idea is that userA is the splitter and userB is the only split receiver.
So we expect in this case all the money in (splittable + collectable of both userA and userB) will not be affected by the split().

Reply:
Then again, the case in which they are equal is interesting to check.
The rule should be correct even then, no?

Reply:
I agree, I have changed the rule to moneyNotLostOrCreatedDuringSplit.
Now there is no require that userA != userB,
so even the case userA == userB is checked.

And the rule is still verified.